### PR TITLE
ICU-20912 Make C/J Currency consistent on lowercase/uppercase currency equality

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20912
- [x] Updated PR title and link in previous line to include Issue number
- [ ] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

**Comment by sffc** [comment]
Ping

**Comment by sffc** [comment]
I just realized there is a depstest error :confused: 

```
Error: in group units: i18n/currunit.o imports u_toupper_66 but units does not depend on ucase (for common/ucase.o)
```

I'll try using only uprv_toupper instead of u_toupper.

**Comment by jefgen** [comment]
> I'll try using only uprv_toupper instead of u_toupper.
Just a thought: Aren't ISO currency codes ASCII/Invariant anyways? (So perhaps it would be good to use uprv_toupper anyways...)

**Comment by sffc** [comment]
Right, except uprv_toupper takes a `char` but currency codes are passed around as `UChar`.  So I will either need to make an overload of uprv_toupper that takes a UChar (Markus suggested naming it uprv_asciiToUpper), or convert back and forth between `char` and `UChar`.

**Comment by jefgen** [comment]
Oh right, good point about the `char` vs `UChar`. Looks like there isn't any uprv_* that does `UChar`.

**Comment by sffc** [comment]
Fixed via amend.

**Comment by sffc** [comment]
Ping

**Comment by sffc** [comment]
I think this PR is complete.  Can someone review and approve please?

File: icu4c/source/test/intltest/numfmtst.cpp [hunk 1]
```diff
@@ -2274,6 +2278,27 @@ void NumberFormatTest::TestCurrencyUnit(void){
...
     assertEquals("Copying should not give uninitialized ISO code", u"", failure.getISOCurrency());
 
+    // Test equality
+    ec = U_ZERO_ERROR;
+    assertFalse("FAIL: USD == JPY", CurrencyUnit(USD, ec) == CurrencyUnit(JPY, ec));
+    assertTrue("FAIL: USD != USD",  CurrencyUnit(USD, ec) == CurrencyUnit(USD, ec));
+    assertTrue("FAIL: JPY != jpy",  CurrencyUnit(JPY, ec) == CurrencyUnit(jpy, ec));
+    assertTrue("FAIL: jpy != JPY",  CurrencyUnit(jpy, ec) == CurrencyUnit(JPY, ec));
+
+    // Test equality with UTF-8 instances
```
Comment (markusicu) [inline icu4c/source/test/intltest/numfmtst.cpp]: Not UTF-8. Say "invariant character" or "system charset" or similar. On z/OS these are in EBCDIC.

File: icu4c/source/test/intltest/numfmtst.cpp [hunk 2]
```diff
@@ -2274,6 +2278,27 @@ void NumberFormatTest::TestCurrencyUnit(void){
...
+    assertFalse("FAIL: USD8 == JPY8", CurrencyUnit(USD8, ec) == CurrencyUnit(JPY8, ec));
+    assertTrue("FAIL: USD8 != USD8",  CurrencyUnit(USD8, ec) == CurrencyUnit(USD8, ec));
+    assertTrue("FAIL: JPY8 != jpy8",  CurrencyUnit(JPY8, ec) == CurrencyUnit(jpy8, ec));
+    assertTrue("FAIL: jpy8 != JPY8",  CurrencyUnit(jpy8, ec) == CurrencyUnit(JPY8, ec));
+
+    // Test equality between UTF-16 and UTF-8 instances
```
Comment (markusicu) [inline icu4c/source/test/intltest/numfmtst.cpp]: ditto

File: icu4c/source/i18n/currunit.cpp [hunk 3]
```diff
@@ -42,10 +44,15 @@ CurrencyUnit::CurrencyUnit(ConstChar16Ptr _isoCode, UErrorCode& ec) {
...
     } else {
-        isoCodeToUse = _isoCode;
+        for (int32_t i=0; i<3; i++) {
+            isoCode[i] = u_asciiToUpper(_isoCode[i]);
+        }
+        isoCode[3] = 0;
+        isoCodeToUse = nullptr;
+    }
+    if (isoCodeToUse != nullptr) {
+        uprv_memcpy(isoCode, isoCodeToUse, sizeof(UChar) * 4);
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/currunit.cpp]: Done; thanks, it's more readable now.

**Comment by sffc** [comment]
Rebased to master and fixed the conflict

**Pull request merged**